korge-fleks/src/commonMain/kotlin/com/github/quillraven/fleks/collection/bitArray.kt

```diff
@@ -0,0 +1,156 @@
+        bits[word] = bits[word] or (1L shl (idx % 64))
+    }
+
+    fun clearAll() {
+        bits.fill(0L)
+    }
+
+    fun clear(idx: Int) {
+        val word = idx / 64
+        if (word < bits.size) {
+            bits[word] = bits[word] and (1L shl (idx % 64)).inv()
+        }
+    }
+
+    fun intersects(other: BitArray): Boolean {
+        val otherBits = other.bits
+        val start = min(bits.size, otherBits.size) - 1
+        for (i in start downTo 0) {
+            if ((bits[i] and otherBits[i]) != 0L) {
+                return true
+            }
+        }
+        return false
+    }
+
+    fun contains(other: BitArray): Boolean {
+        val otherBits = other.bits
+
+        // check if other BitArray is larger and if there is any of those bits set
+        for (i in bits.size until otherBits.size) {
+            if (otherBits[i] != 0L) {
+                return false
+            }
+        }
+
+        // check overlapping bits
+        val start = min(bits.size, otherBits.size) - 1
+        for (i in start downTo 0) {
+            if ((bits[i] and otherBits[i]) != otherBits[i]) {
+                return false
+            }
+        }
+
+        return true
+    }
+
+    /**
+     * Returns the logical size of the [BitArray] which is equal to the highest index of the
+     * bit that is set.
+     *
+     * Returns zero if the [BitArray] is empty.
+     */
+    fun length(): Int {
+        for (word in bits.size - 1 downTo 0) {
+            val bitsAtWord = bits[word]
+            if (bitsAtWord != 0L) {
+                for (bit in 63 downTo 0) {
+                    if ((bitsAtWord and (1L shl (bit % 64))) != 0L) {
+                        return (word shl 6) + bit + 1
+                    }
+                }
+            }
+        }
+        return 0
+    }
+
+    inline fun forEachSetBit(action: (Int) -> Unit) {
+        for (word in bits.size - 1 downTo 0) {
+            val bitsAtWord = bits[word]
+            if (bitsAtWord != 0L) {
+                for (bit in 63 downTo 0) {
+                    if ((bitsAtWord and (1L shl (bit % 64))) != 0L) {
+                        action((word shl 6) + bit)
+                    }
+                }
+            }
+        }
+    }
+
+    fun toIntBag(bag: IntBag) {
+        var checkSize = true
+        bag.clear()
+        forEachSetBit { idx ->
+            if (checkSize) {
+                checkSize = false
+                bag.ensureCapacity(idx)
+            }
+            bag.unsafeAdd(idx)
+        }
+    }
+
+    override fun hashCode(): Int {
+        if (bits.isEmpty()) {
+            return 0
+        }
+
+        val word = length() / 64
+        var hash = 0
+        for (i in 0..word) {
+            hash = 127 * hash + (bits[i] xor (bits[i] ushr 32)).toInt()
+        }
+        return hash
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (other == null) return false
+        if (this === other) return true
+
+        other as BitArray
```

Comment:
This will fail instead of returning false, in the case `other` is from another type, right?

Reply:
I will check :)